### ISSUE TITLE
[BUG] `HierarchicalKMeans.fit()` takes transposed X

### ISSUE THREAD

**Issue by man-shu** [nilearn/nilearn#4755]
.... I noticed that `HierarchicalKMeans.fit()` takes transposed X:
https://github.com/nilearn/nilearn/blob/135e5388386c24ca879328040273127f8acb4a6a/nilearn/regions/hierarchical_kmeans_clustering.py#L233

See in `ReNA.fit()` for comparison:
https://github.com/nilearn/nilearn/blob/135e5388386c24ca879328040273127f8acb4a6a/nilearn/regions/rena_clustering.py#L518

Is this a bug or a feature?

_Originally posted by @man-shu in https://github.com/nilearn/nilearn/pull/4577#discussion_r1843840330_

**Comment by Remi-Gau** [comment]
not had time to check closely: is this a doc issue or an actual bug where the code fails ?

**Comment by man-shu** [comment]
The code does not fail as long as you provide the data the "right" way (i.e. by transposing). See the test for example:
https://github.com/nilearn/nilearn/blob/d73672d41992659395c401fbc08e876708aa78ce/nilearn/regions/tests/test_hierarchical_kmeans_clustering.py#L104

But I think the user should not have to do the transposition unless there is a specific reason to do so (which I might be missing).

It also simply goes against the sklearn convention that we follow...

**Comment by Remi-Gau** [comment]
yeah that is a departure from our API so I would defo file this as bug and push a fix without going through deprecation cycle

**Comment by emdupre** [comment]
> But I think the user should not have to do the transposition unless there is a specific reason to do so (which I might be missing).

Went back to the original PR, and I don't see any discussion on this ; agreed as a bug and push through a fix without deprecation cycles !
